### ISSUE TITLE
Send/Sync impls / missing bounds

### ISSUE THREAD

**Issue by sdroege** [gtk-rs/gir#379]
Currently the assumption seems to be that all GObjects and other data types are not Send and/or Sync, and every callback (e.g. for signals) is only ever called from the same thread where the connection was made to.

This is true in the context of Gtk/Gdk (neither Send nor Sync) or Pango (Send but not Sync), but not for many things in GIO (and even GLib, see GClosure or the GObject::notify signal, both being Send+Sync) or especially GStreamer. For the case of GStreamer the default would be to assume that everything is Send+Sync, and every signal or other callback's closure needs Sync+Send bounds (the number of exceptions can be counted on one hand).

As such it would be good to somehow be able to configure this (as GI does not have any annotations for this):

1. A global per-library configuration for a default. `default_concurrency` with values `none`, `send`, `sync`, `send+sync`. The default should probably stay `none`
2. Per object / signal configuration for the same, which overrides the above. `concurrency`

This would then impl the corresponding traits for the structs, and require the corresponding trait bounds for signals, etc.

The main open question is how to handle this in the context of sub-typing. While e.g. GObject::notify should be Send+Sync in the case of Send+Sync sub-types (e.g. GstElement), it is not required to be so otherwise (e.g. GtkWidget).

Any comments, ideas, ...? :)

**Comment by EPashkin** [comment]
What problem can arise is base function don't have extra constraint and we add it only for function that use it (like signals connect)?

**Comment by sdroege** [comment]
Can you give a concrete example for what you mean?

**Comment by sdroege** [comment]
You mean for example connecting to e.g. the `notify` signal of GObject with a closure without `Send+Sync` for an object that actually is `Send+Sync`? In that case your closure could be called from multiple threads at once, while whatever data you pass into the closure might not be thread-safe at all (e.g. using Rc instead of Arc, not using Mutex but RefCell, ...).

**Comment by EPashkin** [comment]
If glib::signal::connect don't have Send+Sync function constraint,
but we add it for gio::ActionGroupExt::connect_action_added (just for example, don't checked that it really needed), there will be problem?
If not then your proposal is very good.

**Comment by EPashkin** [comment]
It really problem that we can't add limits on part glib::signal::connect call (for connecting "notify::activate" as need for GtkSwitch)

**Comment by sdroege** [comment]
`glib::signal::connect` is unsafe, so that's not really a problem. Its callers will have to set the appropriate bounds. `gtk::Switch::connect_notify_activate` is also not *really* a problem (GtkSwitch being neither `Send` nor `Sync`), and e.g. `gst::Pad::connect_notify_caps` wouldn't be neither (the `Send+Sync` bound could be added there specifically).

The problem is with functions/signals from some base type that can be used verbatim in a sub type. For example if you had a generic `glib::Object::connect_notify` it would be a problem, as for e.g. `gtk::Switch` this wouldn't need the `Send+Sync` bound, but for `gst::Pad` it would need it.

**Comment by EPashkin** [comment]
So current "solution": make all base functions that use closure (like glib::signal::connect)  "unsafe" and add constraint in concrete usage?

**Comment by sdroege** [comment]
`glib::signal::connect` already is unsafe, and has to anyway because you'll have to convert from one type of closure to another in the general case.

I don't think there already is any case yet that is a problem and not marked unsafe, it's just something to keep in mind for the future

**Comment by sdroege** [comment]
@GuillaumeGomez Do you have any opinion on this?

**Comment by GuillaumeGomez** [comment]
My problem with this is that it'll require objects to have one more level in order to avoid data racing issues. I don't know if adding such a thing wouldn't be too complicated or too big at this point.

**Comment by sdroege** [comment]
What do you mean with levels?

**Comment by GuillaumeGomez** [comment]
For now we (generally) only have one level : binding. So gtk-rs is "just" a binding to the C code. If we need to add Send+Sync; we'll have another level. (I hope it's clear enough ^^').

**Comment by sdroege** [comment]
I don't think I understand. Why couldn't you just (correctly) add Sync/Send to the structs/closures as needed? This is not really a different *layer* but something that is even inherent in the C API already (just that there you have it written in English inside the documentation). There's not really any need for a different API or API layer

**Comment by GuillaumeGomez** [comment]
We *can*. My problem is that gtk objects aren't thread safe, which means we need to do add this "protection" ourselves. We can do it "easily" by wrapping the c object inside a mutex, however there is a cost in term of performance, and I don't want this.

We can however target some specific objects to avoid this cost and only do this for the wanted objects. But even for them, if they're mostly used in non-thread context, we'll need to think to add this thread safety *outside* of the basic object.

So to summarize this: I'm all in favor of adding thread-safety as long as it doesn't cost for "normal" usage.

**Comment by sdroege** [comment]
> We can. My problem is that gtk objects aren't thread safe

Yes, *GTK* but not every GLib/GObject based library. E.g. GIO and GStreamer are perfectly thread-safe (Send+Sync) (for most parts), and many non-GTK GObject libraries are at least Send.

I don't want to add thread-safety to GTK by wrapping things in mutexes or things like that (which is not even sufficient for GTK). I just want to add correct trait bounds / impls. That is, for GTK nothing would change (GTK objects are neither Send nor Sync), for other things Send and/or Sync would be added.

**Comment by GuillaumeGomez** [comment]
For objects that are already thread-safe, then no problem. Question then: does gir specify such a thing or not?

**Comment by sdroege** [comment]
No, that's why there needs to be configuration for that as I suggested at the very beginning :)

**Comment by GuillaumeGomez** [comment]
Arf sorry, I focused too much on gtk objects. So about the syntax now that we agree on everything else: I'm not a big fan of "default_concurrency" so I suggest just "concurrency" (tadam!).

**Comment by sdroege** [comment]
Great, works for me :) This and the array length thing are currently my main blockers for GStreamer bindings. Hope I find some time for implementing both this week.

**Comment by GuillaumeGomez** [comment]
If you don't, I'll certainly this week-end.

**Comment by sdroege** [comment]
For array lengths, I have half of an implementation already. This one here with Sync/Send should be rather simple after all
